File: integration/ts/ec-isolation.test.ts
/**
 * EC isolation integration tests — inline contracts testing individual EC functions.
 *
 * Each test compiles a minimal stateless contract that exercises a single EC
 * built-in, deploys it on regtest, and spends via contract.call().
 */

import { describe, it, expect } from 'vitest';
import { compileSource } from './helpers/compile.js';
import { RunarContract } from 'runar-sdk';
import { createFundedWallet } from './helpers/wallet.js';
import { ecMulGen, encodePoint, EC_N, EC_P, EC_GX, EC_GY } from './helpers/crypto.js';
import { createProvider } from './helpers/node.js';

describe('EC Isolation', () => {
  it('ecOnCurve: should compile and deploy a contract checking point validity', async () => {
    const source = `
import { SmartContract, assert, ecOnCurve } from 'runar-lang';
import type { Point } from 'runar-lang';

class EcOnCurveTest extends SmartContract {
  readonly p: Point;

  constructor(p: Point) {
    super(p);
    this.p = p;
  }

  public verify() {
    assert(ecOnCurve(this.p));
  }
}
`;
    const artifact = compileSource(source, 'EcOnCurveTest.runar.ts');
    expect(artifact.contractName).toBe('EcOnCurveTest');

    // Use a valid point on the curve (generator * some scalar)
    const [gx, gy] = ecMulGen(42n);
    const pointHex = encodePoint(gx, gy);

    const contract = new RunarContract(artifact, [pointHex]);

    const provider = createProvider();
    const { signer } = await createFundedWallet(provider);

    const { txid: deployTxid } = await contract.deploy(provider, signer, { satoshis: 5000 });
    expect(deployTxid).toBeTruthy();

    const { txid: spendTxid } = await contract.call('verify', [], provider, signer);
    expect(spendTxid).toBeTruthy();
    expect(spendTxid.length).toBe(64);
  });

  it('ecMulGen: should compile and deploy a scalar-multiply-generator contract', async () => {
    const source = `
import { SmartContract, assert, ecMulGen, ecPointX, ecPointY } from 'runar-lang';
import type { Point } from 'runar-lang';

class EcMulGenTest extends SmartContract {
  readonly expected: Point;

  constructor(expected: Point) {
    super(expected);
    this.expected = expected;
  }

  public verify(k: bigint) {
    const result = ecMulGen(k);
    assert(ecPointX(result) === ecPointX(this.expected));
    assert(ecPointY(result) === ecPointY(this.expected));
  }
}
`;
    const artifact = compileSource(source, 'EcMulGenTest.runar.ts');
    expect(artifact.contractName).toBe('EcMulGenTest');

    const k = 7n;
    const [ex, ey] = ecMulGen(k);
    const expectedHex = encodePoint(ex, ey);

    const contract = new RunarContract(artifact, [expectedHex]);

    const provider = createProvider();
    const { signer } = await createFundedWallet(provider);

    const { txid: deployTxid } = await contract.deploy(provider, signer, { satoshis: 5000 });
    expect(deployTxid).toBeTruthy();

    const { txid: spendTxid } = await contract.call('verify', [k], provider, signer);
    expect(spendTxid).toBeTruthy();
    expect(spendTxid.length).toBe(64);
  });

  it('ecAdd: should compile and deploy a point addition contract', async () => {
    const source = `
import { SmartContract, assert, ecAdd, ecPointX, ecPointY } from 'runar-lang';
import type { Point } from 'runar-lang';

class EcAddTest extends SmartContract {
  readonly a: Point;
  readonly b: Point;
  readonly expected: Point;

  constructor(a: Point, b: Point, expected: Point) {
    super(a, b, expected);
    this.a = a;
    this.b = b;
    this.expected = expected;
  }

  public verify() {
    const result = ecAdd(this.a, this.b);
    assert(ecPointX(result) === ecPointX(this.expected));
    assert(ecPointY(result) === ecPointY(this.expected));
  }
}
`;
    const artifact = compileSource(source, 'EcAddTest.runar.ts');
    expect(artifact.contractName).toBe('EcAddTest');

    const [ax, ay] = ecMulGen(3n);
    const [bx, by] = ecMulGen(5n);
    // 3G + 5G = 8G
    const [ex, ey] = ecMulGen(8n);

    const contract = new RunarContract(artifact, [
      encodePoint(ax, ay),
      encodePoint(bx, by),
      encodePoint(ex, ey),
    ]);

    const provider = createProvider();
    const { signer } = await createFundedWallet(provider);

    const { txid: deployTxid } = await contract.deploy(provider, signer, { satoshis: 5000 });
    expect(deployTxid).toBeTruthy();

    const { txid: spendTxid } = await contract.call('verify', [], provider, signer);
    expect(spendTxid).toBeTruthy();
    expect(spendTxid.length).toBe(64);
  });

  it('ecNegate: should compile, deploy, and spend a point negation contract', async () => {
    const source = `
import { SmartContract, assert, ecNegate, ecPointY } from 'runar-lang';
import type { Point } from 'runar-lang';

class EcNegateTest extends SmartContract {
  readonly pt: Point;

  constructor(pt: Point) {
    super(pt);
    this.pt = pt;
  }

  public check(expectedNegY: bigint) {
    assert(ecPointY(ecNegate(this.pt)) === expectedNegY);
  }
}
`;
    const artifact = compileSource(source, 'EcNegateTest.runar.ts');
    expect(artifact.contractName).toBe('EcNegateTest');

    const [px, py] = ecMulGen(10n);
    const contract = new RunarContract(artifact, [encodePoint(px, py)]);

    const provider = createProvider();
    const { signer } = await createFundedWallet(provider);

    const { txid: deployTxid } = await contract.deploy(provider, signer, { satoshis: 5000 });
    expect(deployTxid).toBeTruthy();

    // ecNegate(P).y === EC_P - P.y on secp256k1
    const negY = EC_P - py;
    const { txid: spendTxid } = await contract.call('check', [negY], provider, signer);
    expect(spendTxid).toBeTruthy();
    expect(spendTxid.length).toBe(64);
  });

  it('ecPointX: should extract the X coordinate of a point', async () => {
    const source = `
import { SmartContract, assert, ecPointX } from 'runar-lang';
import type { Point } from 'runar-lang';

class EcPointXTest extends SmartContract {
  readonly pt: Point;

  constructor(pt: Point) {
    super(pt);
    this.pt = pt;
  }

  public check(expectedX: bigint) {
    assert(ecPointX(this.pt) === expectedX);
  }
}
`;
    const artifact = compileSource(source, 'EcPointXTest.runar.ts');
    const pointHex = encodePoint(EC_GX, EC_GY);
    const contract = new RunarContract(artifact, [pointHex]);

    const provider = createProvider();
    const { signer } = await createFundedWallet(provider);

    await contract.deploy(provider, signer, { satoshis: 5000 });

    const { txid } = await contract.call('check', [EC_GX], provider, signer);
    expect(txid).toBeTruthy();
  });

  it('ecOnCurveThenPointX: should verify point is on curve and extract X', async () => {
    const source = `
import { SmartContract, assert, ecOnCurve, ecPointX } from 'runar-lang';
import type { Point } from 'runar-lang';

class EcOnCurveTwice extends SmartContract {
  readonly pt: Point;

  constructor(pt: Point) {
    super(pt);
    this.pt = pt;
  }

  public check(expectedX: bigint) {
    assert(ecOnCurve(this.pt));
    assert(ecPointX(this.pt) === expectedX);
  }
}
`;
    const artifact = compileSource(source, 'EcOnCurveTwice.runar.ts');
    const pointHex = encodePoint(EC_GX, EC_GY);
    const contract = new RunarContract(artifact, [pointHex]);

    const provider = createProvider();
    const { signer } = await createFundedWallet(provider);

    await contract.deploy(provider, signer, { satoshis: 5000 });

    const { txid } = await contract.call('check', [EC_GX], provider, signer);
    expect(txid).toBeTruthy();
  });

  it('convergencePattern: should verify EC convergence proof', async () => {
    const source = `
import { SmartContract, assert, ecOnCurve, ecAdd, ecNegate, ecMulGen, ecPointX, ecPointY } from 'runar-lang';
import type { Point } from 'runar-lang';

class ConvergencePattern extends SmartContract {
  readonly rA: Point;
  readonly rB: Point;

  constructor(rA: Point, rB: Point) {
    super(rA, rB);
    this.rA = rA;
    this.rB = rB;
  }

  public proveConvergence(deltaO: bigint) {
    assert(ecOnCurve(this.rA));
    assert(ecOnCurve(this.rB));
    const diff = ecAdd(this.rA, ecNegate(this.rB));
    const expected = ecMulGen(deltaO);
    assert(ecPointX(diff) === ecPointX(expected));
    assert(ecPointY(diff) === ecPointY(expected));
  }
}
`;
    const artifact = compileSource(source, 'ConvergencePattern.runar.ts');
    const a = 142n;
    const b = 37n;
    const deltaO = a - b;
    const [rAx, rAy] = ecMulGen(a);
    const [rBx, rBy] = ecMulGen(b);

    const contract = new RunarContract(artifact, [
      encodePoint(rAx, rAy),
      encodePoint(rBx, rBy),
    ]);

    const provider = createProvider();
    const { signer } = await createFundedWallet(provider);

    await contract.deploy(provider, signer, { satoshis: 500000 });

    const { txid } = await contract.call('proveConvergence', [deltaO], provider, signer);
    expect(txid).toBeTruthy();
  });

  it('ecMulGen with large scalar: should handle scalars near curve order', async () => {
    const source = `
import { SmartContract, assert, ecMulGen, ecPointX, ecPointY } from 'runar-lang';
import type { Point } from 'runar-lang';

class EcMulGenTest extends SmartContract {
  readonly expected: Point;

  constructor(expected: Point) {
    super(expected);
    this.expected = expected;
  }

  public verify(k: bigint) {
    const result = ecMulGen(k);
    assert(ecPointX(result) === ecPointX(this.expected));
    assert(ecPointY(result) === ecPointY(this.expected));
  }
}
`;
    const artifact = compileSource(source, 'EcMulGenTest.runar.ts');
    const k = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364100n;
    const [ex, ey] = ecMulGen(k);
    const contract = new RunarContract(artifact, [encodePoint(ex, ey)]);

    const provider = createProvider();
    const { signer } = await createFundedWallet(provider);

    await contract.deploy(provider, signer, { satoshis: 5000 });

    const { txid } = await contract.call('verify', [k], provider, signer);
    expect(txid).toBeTruthy();
  });
});

File: integration/zig/src/ec_isolation_test.zig
const std = @import("std");
const runar = @import("runar");
const helpers = @import("helpers.zig");
const compile = @import("compile.zig");

// ---------------------------------------------------------------------------
// EC isolation tests: verify individual EC operations compile and deploy
// correctly. These test the EC demo contract which exercises ecOnCurve,
// ecPointX, ecPointY, ecAdd, ecNegate, ecMulGen, etc.
// ---------------------------------------------------------------------------

test "ECIsolation_Compile_ECDemo" {
    const allocator = std.testing.allocator;

    var artifact = compile.compileContract(allocator, "examples/zig/ec-demo/ECDemo.runar.zig") catch |err| {
        std.log.warn("Could not compile ECDemo contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    try std.testing.expectEqualStrings("ECDemo", artifact.contract_name);
    std.log.info("ECDemo compiled: {d} bytes", .{artifact.script.len / 2});
}

test "ECIsolation_ScriptSize" {
    const allocator = std.testing.allocator;

    var artifact = compile.compileContract(allocator, "examples/zig/ec-demo/ECDemo.runar.zig") catch |err| {
        std.log.warn("Could not compile ECDemo contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    const script_bytes = artifact.script.len / 2;
    // EC contracts produce large scripts due to field arithmetic
    try std.testing.expect(script_bytes > 1000);
    std.log.info("ECDemo script size: {d} bytes", .{script_bytes});
}

test "ECIsolation_Deploy_ECDemo" {
    const allocator = std.testing.allocator;

    if (!helpers.isNodeAvailable(allocator)) {
        std.log.warn("Regtest node not available, skipping test", .{});
        return;
    }

    var artifact = compile.compileContract(allocator, "examples/zig/ec-demo/ECDemo.runar.zig") catch |err| {
        std.log.warn("Could not compile ECDemo contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    // secp256k1 generator point G as 64-byte hex (x[32]||y[32])
    const g_point_hex = "79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798" ++
        "483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8";

    var contract = try runar.RunarContract.init(allocator, &artifact, &[_]runar.StateValue{
        .{ .bytes = g_point_hex },
    });
    defer contract.deinit();

    var wallet = try helpers.newWallet(allocator);
    defer wallet.deinit();
    const fund_txid = try helpers.fundWallet(allocator, &wallet, 1.0);
    defer allocator.free(fund_txid);

    var rpc_provider = helpers.RPCProvider.init(allocator);
    var local_signer = try wallet.localSigner();

    const deploy_txid = try contract.deploy(rpc_provider.provider(), local_signer.signer(), .{ .satoshis = 500000 });
    defer allocator.free(deploy_txid);

    try std.testing.expectEqual(@as(usize, 64), deploy_txid.len);
    std.log.info("ECDemo deployed: {s}", .{deploy_txid});
}

test "ECIsolation_Compile_ConvergenceProof" {
    const allocator = std.testing.allocator;

    // The ConvergenceProof contract uses multiple EC operations together:
    // ecOnCurve, ecAdd, ecNegate, ecMulGen, ecPointX, ecPointY
    var artifact = compile.compileContract(allocator, "examples/zig/convergence-proof/ConvergenceProof.runar.zig") catch |err| {
        std.log.warn("Could not compile ConvergenceProof contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    try std.testing.expectEqualStrings("ConvergenceProof", artifact.contract_name);
    std.log.info("ConvergenceProof compiled: {d} bytes (EC operations)", .{artifact.script.len / 2});
}

test "ECIsolation_Deploy_DifferentPoints" {
    const allocator = std.testing.allocator;

    if (!helpers.isNodeAvailable(allocator)) {
        std.log.warn("Regtest node not available, skipping test", .{});
        return;
    }

    var artifact = compile.compileContract(allocator, "examples/zig/ec-demo/ECDemo.runar.zig") catch |err| {
        std.log.warn("Could not compile ECDemo contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    // Deploy with G point
    const g_point_hex = "79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798" ++
        "483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8";

    var contract1 = try runar.RunarContract.init(allocator, &artifact, &[_]runar.StateValue{
        .{ .bytes = g_point_hex },
    });
    defer contract1.deinit();

    var wallet1 = try helpers.newWallet(allocator);
    defer wallet1.deinit();
    const fund_txid1 = try helpers.fundWallet(allocator, &wallet1, 1.0);
    defer allocator.free(fund_txid1);

    var rpc1 = helpers.RPCProvider.init(allocator);
    var signer1 = try wallet1.localSigner();
    const txid1 = try contract1.deploy(rpc1.provider(), signer1.signer(), .{ .satoshis = 500000 });
    defer allocator.free(txid1);

    // Deploy with 2*G point
    const two_g_hex = "c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5" ++
        "1ae168fea63dc339a3c58419466ceae1032688d15f9c819fea738c882b9d5d90";

    var contract2 = try runar.RunarContract.init(allocator, &artifact, &[_]runar.StateValue{
        .{ .bytes = two_g_hex },
    });
    defer contract2.deinit();

    var wallet2 = try helpers.newWallet(allocator);
    defer wallet2.deinit();
    const fund_txid2 = try helpers.fundWallet(allocator, &wallet2, 1.0);
    defer allocator.free(fund_txid2);

    var rpc2 = helpers.RPCProvider.init(allocator);
    var signer2 = try wallet2.localSigner();
    const txid2 = try contract2.deploy(rpc2.provider(), signer2.signer(), .{ .satoshis = 500000 });
    defer allocator.free(txid2);

    // Different constructor args produce different scripts
    try std.testing.expect(!std.mem.eql(u8, txid1, txid2));
    std.log.info("ECDemo G: {s}, 2G: {s}", .{ txid1, txid2 });
}
